File: articles/design-thinking-public-policy/go/policy_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type PolicyPilot struct {
	Name            string
	Accessibility   float64
	Feasibility     float64
	Legitimacy      float64
	Equity          float64
	BurdenReduction float64
	Durability      float64
	Risk            float64
	PolicyValue     float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func policyValue(x PolicyPilot) float64 {
	return 0.20*x.Accessibility +
		0.16*x.Feasibility +
		0.16*x.Legitimacy +
		0.20*x.Equity +
		0.14*x.BurdenReduction +
		0.08*x.Durability -
		0.06*x.Risk
}

func loadPilots(path string) []PolicyPilot {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []PolicyPilot

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 10 {
			log.Fatalf("malformed row: %v", row)
		}

		x := PolicyPilot{
			Name:            row[0],
			Accessibility:   parseFloat(row[3]),
			Feasibility:     parseFloat(row[4]),
			Legitimacy:      parseFloat(row[5]),
			Equity:          parseFloat(row[6]),
			BurdenReduction: parseFloat(row[7]),
			Durability:      parseFloat(row[8]),
			Risk:            parseFloat(row[9]),
		}
		x.PolicyValue = policyValue(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/public_policy_pilots_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadPilots(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].PolicyValue > items[j].PolicyValue
	})

	fmt.Println("rank,pilot,policy_value,risk")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.PolicyValue, x.Risk)
	}
}

File: articles/design-thinking-public-policy/cpp/policy_value_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct PolicyPilot {
    std::string name;
    double accessibility;
    double feasibility;
    double legitimacy;
    double equity;
    double burden_reduction;
    double durability;
    double risk;
};

struct ScoredPilot {
    std::string name;
    double value;
    double risk;
};

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }
    return fields;
}

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

double policy_value(const PolicyPilot& x) {
    return 0.20 * x.accessibility +
           0.16 * x.feasibility +
           0.16 * x.legitimacy +
           0.20 * x.equity +
           0.14 * x.burden_reduction +
           0.08 * x.durability -
           0.06 * x.risk;
}

std::vector<PolicyPilot> read_pilots(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<PolicyPilot> pilots;
    std::string line;
    std::getline(file, line); // header

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 10) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        pilots.push_back(
            PolicyPilot{
                fields[0],
                to_double(fields[3]),
                to_double(fields[4]),
                to_double(fields[5]),
                to_double(fields[6]),
                to_double(fields[7]),
                to_double(fields[8]),
                to_double(fields[9])
            }
        );
    }

    return pilots;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/public_policy_pilots_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto pilots = read_pilots(input_path);
        std::vector<ScoredPilot> scored;

        for (const auto& pilot : pilots) {
            scored.push_back(
                ScoredPilot{
                    pilot.name,
                    policy_value(pilot),
                    pilot.risk
                }
            );
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredPilot& a, const ScoredPilot& b) {
                return a.value > b.value;
            }
        );

        std::cout << "rank,pilot,policy_value,risk\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].value << ","
                      << scored[i].risk << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
